Add tcp_wrappers and vsftpd actors

olysonek opened the pull request:
This pull request adds a couple of actors and some helpers for testing.

The `TcpWrappersConfigRead` actor reads the tcp_wrappers configuration (/etc/hosts.{allow,deny}) and stores some data about it in the `TcpWrappersFacts` model. An instance of that model can then be passed to the `config_applies_to_daemon` library function which can be used in other actors to determine, if the tcp_wrappers configuration applies to the specified daemon.

The `VsftpdConfigRead` actor gathers some data about the vsftpd configuration (but only if vsftpd is installed) and  stores it in the `VsftpdFacts` model.

The `VsftpdConfigCheck` actor consumes `TcpWrappersFacts` and `VsftpdFacts`. The actor determines, whether the vsftpd configuration is supported in RHEL-8. If it is not, it inhibits the upgrade. Namely, it checks whether vsftpd was using tcp_wrappers, which is not supported in RHEL-8.

The `VsftpdConfigUpdate` actor consumes `VsftpdFacts` and modifies the vsftpd config files so that they still work in RHEL-8, and so that the effective configuration is as close to the original as possible. See the comit message of the last commit for more details.

fernflower commented on repos/system_upgrade/el7toel8/libraries/tests/test_vsftpdutils.py:
```diff
@@ -0,0 +1,82 @@
+
+def test_getting_inaccessible_config_gives_None():
+    path = 'my_file'
+    f = MockFile(path, to_raise=IOError(errno.EACCES))
```
Nit: 
While having read_file function (leapp.libraries.common.vsftpdutils.read_file) implemented separately makes the read file functionality easily testable, having it passed as a callback into other methods like get_config_contents and get_default_config_hash imho is overcomplicating the api (honestly I can't think of a second way to read the file).
Have you considered using mock's context manager to control output of open() function instead of doing all the work yourself? 

Something like 
```
with mock.patch('builtins.open', side_effect=IOError) as open_mock:
    res = get_config_contents(path, read_func=f.read_file)
```

It's an open question, I might not have the full picture as I've definitely spent a lot less time with this patch than you had.

olysonek replied:
Personally, I like the idea of injecting dependencies, which interact with the external system. It seems cleaner than making the test reliant on the way the implementation reads the files (another way to open the file would be `os.open`). But I do agree that my implementation of dependency injection is not ideal and makes it look a bit messy. I haven't found a better way so far.

Anyway, I did try mocking `builtins.open`, but couldn't get it to work. It seemed to me that apart from the function under test, something else was also using `open`, so mocking it broke that "something". But I may be wrong.

pirat89 replied:
@olysonek: `builtins.open` is kind of problematic. I thought it is just issue when you want to mock stuff and runnning whole actor in tests. But nowadays we have the `api` library, it is possible that there is hidden use of `open` as well in background.

fernflower commented on repos/system_upgrade/el7toel8/actors/tcpwrappersconfigread/libraries/library.py:
```diff
@@ -0,0 +1,67 @@
+        return f.read()
+
+
+def get_lines(content):
```
Wondering if this will do the trick

```
re.sub(r'//[\S\t\n]*\n', '', content)
return content.split('\n')
```

olysonek replied:
You mean this, right? Good thinking, I'll change it.
```
content = re.sub(r'\\\n', '', content)
return content.split('\n')
```

olysonek commented on repos/system_upgrade/el7toel8/libraries/tcpwrappersutils.py:
```diff
@@ -0,0 +1,81 @@
+    should inhibit the upgrade, so that the upgraded system is not insecure.
+    '''
+    for daemon_list in facts.daemon_lists:
+        value = daemon_list.value
```
> L77-L78 can be squashed
> 
> value = [item.lower() for item in daemon_list.value]

Done.

> Nit: iiuc L76-L81 of config_applies_to_daemon are equivalent to
> 
> return any((_daemon_list_matches_daemon([item.lower() for item in daemon_list.value], daemon, 0) for daemon_list in facts.daemon_lists))

The old variant is more readable to me :).

olysonek commented on repos/system_upgrade/el7toel8/libraries/tests/test_vsftpdutils.py:
```diff
@@ -0,0 +1,82 @@
+        if path != self.path:
+            self.error = True
+            raise ValueError
+        if self.to_raise is None:
```
OK, done.

olysonek commented on repos/system_upgrade/el7toel8/libraries/vsftpdutils.py:
```diff
@@ -0,0 +1,32 @@
+    except IOError as e:
+        if e.errno != errno.ENOENT:
+            api.current_logger().warn('Failed to read vsftpd configuration file: %s' % e)
+        return None
```
I'm not sure what you mean. Can you elaborate?

fernflower replied:
I was saying that the function will return None in any case, with or without explicit return None at L23.
But as you wish, it's not important thing.

olysonek replied:
Well there's two cases:
1. if `read_func` succeeds, i.e. does not raise an exception, `get_config_contents` will return the file contents (it cannot return None in this case)
2. if `read_func` fails, i.e. raises an exception, it means the file could not be read (for whatever reason) and `get_config_contents` returns None

I think all is good here.

olysonek commented on repos/system_upgrade/el7toel8/libraries/vsftpdutils.py:
```diff
@@ -0,0 +1,32 @@
+
+def get_default_config_hash(read_func=read_file):
+    content = get_config_contents(VSFTPD_DEFAULT_CONFIG_PATH, read_func=read_func)
+    if content is None:
```
That would change the behaviour slightly, because `bool("") == False`. I can drop `None` from  `return None` though, if you like.

olysonek commented on repos/system_upgrade/el7toel8/models/vsftpdfacts.py:
```diff
@@ -0,0 +1,17 @@
+
+    path = fields.String()
+    strict_ssl_read_eof = fields.Nullable(fields.Boolean())
+    tcp_wrappers = fields.Nullable(fields.Boolean())
```
If `tcp_wrappers is None`, it means that the "tcp_wrappers" option was left unspecified in the configuration file, so vsftpd will use the default value for the option. We could assign the default value to the attribute, if the option is unspecified in the config file, however I think it's better the way I designed it - this way the config reader is more generic, because reading the config, and deciding what the default value is, is separated.

pirat89 replied:
Discussed in person. Nullable makes sense. will be documented.

olysonek commented on repos/system_upgrade/el7toel8/actors/vsftpdconfigcheck/tests/test_library.py:
```diff
@@ -0,0 +1,59 @@
+
+def test_actor_with_unsupported_tcpwrap_config(current_actor_context):
+    config1 = VsftpdConfig(path='/etc/vsftpd/foo.conf', tcp_wrappers=False)
+    config2 = VsftpdConfig(path='/etc/vsftpd/bar.conf', tcp_wrappers=None)
```
It does, it means the option is unspecified in the configuration file, so the default value was being used. The default value in RHEL-7 is "YES", so if the option in the config file is unspecified, tcp_wrappers support is enabled. In this test case, we test that the actor produces the warning report in such a case.

olysonek replied:
Waaaait a minute. I think the test is wrong and there's a bug...

olysonek replied:
No, it's all good. I was just wrong about what the default value is. The default is "NO".

olysonek commented on repos/system_upgrade/el7toel8/actors/vsftpdconfigread/libraries/library.py:
```diff
@@ -0,0 +1,45 @@
+    for path, config in configs:
+        res_configs.append(VsftpdConfig(path=path,
+                                        strict_ssl_read_eof=config.get(STRICT_SSL_READ_EOF),
+                                        tcp_wrappers=config.get(TCP_WRAPPERS)))
```
We could do that, but the default value would have to be `True`. However, the actor is more generic, if we put the code that decides what the default is, in the config check and config update actors.

olysonek replied:
> We could do that, but the default value would have to be `True`

I take that back. `False` is correct, as that is the default value of the option.

pirat89 replied:
Discussed in person. The `None` makes sense.

olysonek commented on repos/system_upgrade/el7toel8/actors/vsftpdconfigread/libraries/library.py:
```diff
@@ -0,0 +1,45 @@
+    res_configs = []
+    for path, config in configs:
+        res_configs.append(VsftpdConfig(path=path,
+                                        strict_ssl_read_eof=config.get(STRICT_SSL_READ_EOF),
```
Yes, it was.

pirat89 replied:
Discussed in person. The `None` makes sense.

olysonek commented on repos/system_upgrade/el7toel8/actors/vsftpdconfigcheck/libraries/library.py:
```diff
@@ -0,0 +1,19 @@
+                                'there is some vsftpd-related configuration in /etc/hosts.deny or '
+                                '/etc/hosts.allow. Please migrate it manually (see '
+                                'https://access.redhat.com/solutions/3906701.')
+                               % ". ".join(bad_configs),
```
Applied, with some minor modifications. Thanks!

olysonek commented on repos/system_upgrade/el7toel8/actors/vsftpdconfigcheck/libraries/library.py:
```diff
@@ -0,0 +1,19 @@
+        if config.tcp_wrappers:
+            bad_configs.append(config.path)
+    if len(bad_configs) > 0 and config_applies_to_daemon(tcpwrap_facts, 'vsftpd'):
+        report_generic(title='Unsupported vsftpd configuration',
```
Done, thanks.

pirat89 commented on repos/system_upgrade/el7toel8/libraries/vsftpdutils.py:
```diff
@@ -0,0 +1,32 @@
+        return f.read()
+
+
+def get_config_contents(path, read_func=read_file):
```
```suggestion
def _get_config_contents(path, read_func=read_file):
```

olysonek replied:
This turned out to be a public function :). I'll document it. I've renamed the other private functions though.

olysonek replied:
I've documented it.

pirat89 commented on repos/system_upgrade/el7toel8/libraries/vsftpdutils.py:
```diff
@@ -0,0 +1,32 @@
+TCP_WRAPPERS = 'tcp_wrappers'
+
+
+def read_file(path):
```
```suggestion
def _read_file(path):
```

olysonek replied:
Actually, this is public API as well :). I've documented it.

pirat89 commented:
@olysonek This is how you can document fields in models: https://github.com/oamg/leapp/blob/1f842237752860d395e184b4934f8c3ff98dff57/leapp/workflows/__init__.py#L51

It should be the same as we use for classes.

fernflower commented on repos/system_upgrade/el7toel8/actors/tcpwrappersconfigread/libraries/library.py:
```diff
@@ -0,0 +1,57 @@
+    daemon_list = line.split(':', 1)[0]
+    daemon_list = daemon_list.replace(',', ' ')
+    daemon_list = daemon_list.replace('\t', ' ')
+    daemon_list = daemon_list.split(' ')
```
nit: wondering if re.split(',| |\t', daemon_list) is more readable

olysonek replied:
Nice, thanks! I didn't realize that exists :).

olysonek replied:
Done. I've also extended the test coverage a bit.

fernflower commented on repos/system_upgrade/el7toel8/actors/tcpwrappersconfigread/libraries/library.py:
```diff
@@ -0,0 +1,57 @@
+            api.current_logger().warning('Failed to read %s: %s' % (path, e))
+        return ret
+    lines = get_lines(content)
+    lines = [line for line in lines if not is_comment(line)]
```
nit: Squash L41-L45?
```
return [get_daemon_list_in_line(line) for line in get_lines(content) if not is_comment(line)]
```

olysonek replied:
OK. I've learned to instinctively avoid complex list comprehensions, that's why you're seeing this :). But this is just fine. Done.

fernflower commented on repos/system_upgrade/el7toel8/actors/vsftpdconfigcheck/libraries/library.py:
```diff
@@ -0,0 +1,23 @@
+
+
+def check_config_supported(tcpwrap_facts, vsftpd_facts):
+    bad_configs = []
```
nit: I'd rather use list comprehension here as well instead of L6-L9
```bad_configs = [config.path for config in vsftpd_facts.configs if config.tcp_wrappers]```

olysonek replied:
Absolutely, thanks. Done. I've also extended the test coverage a bit.

fernflower commented on repos/system_upgrade/el7toel8/libraries/tcpwrappersutils.py:
```diff
@@ -0,0 +1,80 @@
+        regex += re.escape(pattern[part_beginning:part_end])
+
+        if part_end < len(pattern):
+            if pattern[part_end] == '*':
```
nit: Squash L17-L20 to ternary operator?
```regex += '.*' if pattern[part_end] == '*' else '.'```

olysonek replied:
I find my version more readable.

pirat89 commented on repos/system_upgrade/el7toel8/actors/vsftpdconfigupdate/actor.py:
```diff
@@ -0,0 +1,26 @@
+       with its default content (i.e., unmodified) on the source system (the configuration
+       file gets replaced with a new version during the RPM upgrade in this case).
+    2. Adding 'option=old_effective_value' to configuration files for options whose default
+       default value has changed, if the option is not explicitly specified in the configuration
```
double *default*
```suggestion
       value has changed, if the option is not explicitly specified in the configuration
```

olysonek replied:
Good catch, thanks. Fixed.

pirat89 commented:
Missing a check whether the RPM is installed. Upgrade fails on system without vsftpd:
```
Traceback (most recent call last):
  File "/usr/lib64/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python2.7/site-packages/leapp/repository/actor_definition.py", line 59, in _do_run
    target_actor(logger=logger, messaging=messaging).run(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/leapp/actors/__init__.py", line 299, in run
    self.process(*args)
  File "/usr/share/leapp-repository/repositories/system_upgrade/el7toel8/actors/vsftpdconfigread/actor.py", line 18, in process
    self.produce(library.get_vsftpd_facts())
  File "/usr/share/leapp-repository/repositories/system_upgrade/el7toel8/actors/vsftpdconfigread/libraries/library.py", line 39, in get_vsftpd_facts
    configs = _get_parsed_configs(read_func=read_func, listdir=listdir)
  File "/usr/share/leapp-repository/repositories/system_upgrade/el7toel8/actors/vsftpdconfigread/libraries/library.py", line 24, in _get_parsed_configs
    for fname in listdir(VSFTPD_CONFIG_DIR):
OSError: [Errno 2] No such file or directory: '/etc/vsftpd'
```

olysonek commented:
> Missing a check whether the RPM is installed. Upgrade fails on system without vsftpd:

Thanks, and sorry about that. I've fixed the traceback and added a test for it. I'll add the RPM check too.

pirat89 commented on repos/system_upgrade/el7toel8/actors/vsftpdconfigread/libraries/library.py:
```diff
@@ -0,0 +1,51 @@
+            parsed = _parse_config(path, content)
+            if parsed is not None:
+                res.append((path, parsed))
+    except IOError as e:
```
OSError: [Errno 2] No such file or directory: '/etc/vsftpd'

```suggestion
    except OSError as e:
```

olysonek replied:
Thanks again. Hopefully fixed now. I also noticed that I wasn't creating IOError and OSError in the tests correctly (the errno attribute wasn't being set), so I fixed that. And rebased against origin/master.

olysonek replied:
Notice that there's a newly added first commit.

pirat89 commented:
@olysonek I will wait for the RPM check. anyway, when you add that, please rebase against current master branch. From my light testing when `vsftpd` is installed, everything seems ok and working regarding described behaviour.

olysonek commented:
@fernflower @pirat89 Thanks for the thorough review and testing!

Jakuje commented:
I am still wondering whether there should not be some kind of global check that could inhibit the update whenever there is _anything_ useful in tcp_wrappers files, rather than writing actors per daemon to check this. As far as I know, most of the other daemons really do not need special treating since they do not need another configuration option to enable tcp wrappers like in vsftpd.

For example, I would certainly like to have openssh check, since that is really common, but it would be a very simple one using the library provided by this PR.

Please, let me know about your thoughts what would be preferred.

I would personally like to avoid writing some whitelist of daemons (even though I have some list from fedora [1]) that would be tried since there might be some customer application that used to use the tcp_wrappers and they still think they depend on it even after the rebuild (with a good configure check disabling the support in rebuild) the no longer do.

[1] https://fedoraproject.org/wiki/Changes/Deprecate_TCP_wrappers

olysonek commented:
> I would personally like to avoid writing some whitelist of daemons (even though I have some list from fedora [1]) that would be tried since there might be some customer application that used to use the tcp_wrappers and they still think they depend on it even after the rebuild (with a good configure check disabling the support in rebuild) the no longer do.

That is a good point. We may need to rethink this.

The reason I originally decided to do the check directly in the vsftpd actor is so that the upgrade process is as smooth as possible. I.e. do the reporting and upgrade inhibiting only when really necessary.

For example, let's say there's some tcp_wrappers configuration that applies to vsftpd, but
a) tcp_wrappers support is turned off in the vsftpd configuration, or
b) vsftpd is not even installed (anymore)

In either of these cases, it seems unnecessary to nag the user about unsupported configuration. So I check these conditions in the vsftpd actors.

But given the point you made above about customer applications depending on tcp_wrappers, I'm starting to think a global check might be a better solution after all.

Jakuje commented:
On the other hand, I am quite not sure what assurances we can provide for such custom applications. The tcp wrappers package is gone from RHEL8 so we might even just do a list of applications that used tcp_wrappers in RHEL7 and do this in one separate actor (probably also with check whether the separate package is installed to avoid too much noise).

Please, let me know your thoughts or whether you will have time to check that or I should have a look into that.

pirat89 commented:
@olysonek There is chance that users are using custom / 3rd-party applications that depends on tcp_wrappers itself. From that point, not sure whether we should inhibit the upgrade or just inform users about that. Probably better would be to wait for implementation of the answerfile - so it will be possible to inhibit upgrade by default until user do changes or keep users to decide whether this is OK or not.

olysonek commented:
@Jakuje Please look into it, I'm afraid I won't have the time. I'm happy to do changes to the vsftpd actors however, if that becomes necessary.

@pirat89 The VsftpdConfigCheck actor currently inhibits the upgrade and tells the user to migrate the tcp_wrappers configuration (all that is really needed is to remove the tcp_wrappers configuration). It does this so that the upgraded system is not insecure. Do you think it should not inhibit the upgrade? I mean, we can change it to use answerfile once it becomes available, however until then, personally I'd leave it as it is.
